// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseApis = parseApis;
const util_1 = require("@sap-cloud-sdk/util");
const openapi_types_1 = require("../openapi-types");
const extensions_1 = require("../extensions");
const operation_1 = require("./operation");
const operation_naming_1 = require("./operation-naming");
const unique_naming_1 = require("./unique-naming");
/**
 * Collect and parse all APIs of an `OpenAPIV3.Document`.
 * @param document - The OpenAPI document to parse.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Parser options.
 * @returns A flat list of parsed APIs.
 * @internal
 */
function parseApis(document, refs, options) {
    const operationsByApis = getOperationsByApis(document);
    return Object.entries(operationsByApis).map(([name, operations]) => {
        const namedOperations = (0, operation_naming_1.nameOperations)(operations);
        const operationNames = namedOperations.map(
        // All operations have been named in the previous step
        ({ operation }) => operation.operationId);
        const uniqueNames = (0, unique_naming_1.ensureUniqueNames)(operationNames, options, {
            separator: '_'
        });
        const uniquelyNamedOperations = namedOperations.map((operationInfo, i) => {
            operationInfo.operation.operationId = uniqueNames[i];
            return operationInfo;
        });
        return {
            name,
            operations: uniquelyNamedOperations.map(operationInfo => (0, operation_1.parseOperation)(operationInfo, refs, options))
        };
    });
}
function getPathItem(document, pathPattern) {
    const pathItem = document.paths[pathPattern];
    if (!pathItem) {
        // This should never happen
        throw new Error(`Could not parse APIs. Path pattern '${pathPattern}' does not exist in the document.`);
    }
    return pathItem;
}
function getApiNameForOperation({ operation, pathPattern }, document) {
    var _a;
    const originalApiName = operation[extensions_1.apiNameExtension] ||
        getPathItem(document, pathPattern)[extensions_1.apiNameExtension] ||
        document[extensions_1.apiNameExtension] ||
        ((_a = operation.tags) === null || _a === void 0 ? void 0 : _a[0]) ||
        extensions_1.defaultApiName;
    return `${(0, util_1.pascalCase)(originalApiName.replace(/api$/i, ''))}Api`;
}
function getOperationsByApis(document) {
    const allOperations = getAllOperations(document);
    if (!allOperations.length) {
        throw new Error('Could not parse APIs. The document does not contain any operations.');
    }
    return allOperations.reduce((apiMap, operationInfo) => {
        const apiName = getApiNameForOperation(operationInfo, document);
        if (!apiMap[apiName]) {
            apiMap[apiName] = [];
        }
        apiMap[apiName].push(operationInfo);
        return apiMap;
    }, {});
}
function getAllOperations(openApiDocument) {
    return (0, util_1.flat)(Object.entries(openApiDocument.paths).map(([pathPattern, pathItem]) => openapi_types_1.methods
        .filter(method => pathItem[method])
        .map(method => ({
        pathPattern,
        // We can assume that the operation exists as non existing operations where filtered before.
        operation: pathItem[method],
        pathItemParameters: pathItem.parameters || [],
        method
    }))));
}
