// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/document.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseOpenApiDocument = parseOpenApiDocument;
exports.parseSchemas = parseSchemas;
const swagger_parser_1 = __importDefault(require("@apidevtools/swagger-parser"));
const schema_util_1 = require("../schema-util");
const schema_1 = require("./schema");
const api_1 = require("./api");
const refs_1 = require("./refs");
/**
 * Parse an OpenAPI document.
 * @param fileContent - Original OpenAPI document object.
 * @param serviceOptions - Service options as defined in the options per service.
 * @param options - Parser options.
 * @returns The parsed OpenAPI document representation
 * @internal
 */
async function parseOpenApiDocument(fileContent, serviceOptions, options) {
    const clonedContent = JSON.parse(JSON.stringify(fileContent));
    const document = (await swagger_parser_1.default.parse(clonedContent));
    const refs = await (0, refs_1.createRefs)(document, options);
    const schemas = parseSchemas(document, refs, options);
    sanitizeDiscriminatedSchemas(schemas, refs, options);
    return {
        apis: (0, api_1.parseApis)(document, refs, options),
        serviceDescription: document.info.description,
        serviceOptions,
        serviceName: serviceOptions.directoryName,
        schemas
    };
}
// Some specs include incorrect discriminator definitions based on the schema type `object`, that circularly reference their parent type.
async function sanitizeDiscriminatedSchemas(schemas, refs, options) {
    const discriminatorSchemas = schemas
        .filter(({ schema }) => (0, schema_util_1.isOneOfSchema)(schema) && schema.discriminator)
        // type is known because of the filter above
        .map(({ schema, schemaName }) => ({
        children: Object.values(schema.discriminator.mapping),
        schemaName,
        propertyName: schema.discriminator.propertyName
    }));
    discriminatorSchemas.forEach(discriminatorSchema => {
        discriminatorSchema.children.forEach(childRef => {
            const child = schemas.find(schema => schema.schemaName === childRef.schemaName);
            if ((0, schema_util_1.isAllOfSchema)(child === null || child === void 0 ? void 0 : child.schema)) {
                child.schema.allOf = child.schema.allOf.map(grandChild => {
                    // if grandChild is the parent, aka. circular reference
                    if ((0, schema_util_1.isReferenceObject)(grandChild) &&
                        grandChild.schemaName === discriminatorSchema.schemaName) {
                        const { properties = {}, required } = refs.resolveObject(grandChild);
                        properties[discriminatorSchema.propertyName] = { type: 'string' };
                        return (0, schema_1.parseObjectSchema)({ properties, required, additionalProperties: false }, refs, options);
                    }
                    return grandChild;
                });
            }
        });
    });
}
/**
 * @internal
 */
function parseSchemas(document, refs, options) {
    var _a;
    return Object.entries(((_a = document.components) === null || _a === void 0 ? void 0 : _a.schemas) || {}).map(([name, schema]) => (Object.assign(Object.assign({}, refs.getSchemaNaming(`#/components/schemas/${name}`)), { schema: (0, schema_1.parseSchema)(schema, refs, options), description: refs.resolveObject(schema).description, schemaProperties: (0, schema_1.parseSchemaProperties)(schema), nullable: 'nullable' in schema && schema.nullable ? true : false })));
}
