// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseOperation = parseOperation;
exports.getRelevantParameters = getRelevantParameters;
exports.parsePathPattern = parsePathPattern;
exports.parsePathParameters = parsePathParameters;
exports.parseParameters = parseParameters;
const util_1 = require("@sap-cloud-sdk/util");
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
const request_body_1 = require("./request-body");
const schema_1 = require("./schema");
const responses_1 = require("./responses");
const unique_naming_1 = require("./unique-naming");
/**
 * Parse an operation info into a serialization-ready object.
 * @param operationInfo - Parsing relevant information on an operation.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Parser options.
 * @returns A flat list of parsed operations.
 * @internal
 */
function parseOperation({ operation, pathPattern, method, pathItemParameters }, refs, options) {
    const requestBody = (0, request_body_1.parseRequestBody)(operation.requestBody, refs, options);
    const response = (0, responses_1.parseResponses)(operation.responses, refs, options);
    const relevantParameters = getRelevantParameters([...(pathItemParameters || []), ...(operation.parameters || [])], refs);
    const pathParams = relevantParameters.filter(parameter => parameter.in === 'path');
    const headerParams = relevantParameters.filter(parameter => parameter.in === 'header');
    const queryParams = relevantParameters.filter(parameter => parameter.in === 'query');
    const pathParameters = parsePathParameters(pathParams, refs, options);
    return Object.assign(Object.assign({}, operation), { method,
        requestBody,
        response, queryParameters: parseParameters(queryParams, refs, options), headerParameters: parseParameters(headerParams, refs, options), pathParameters, pathPattern: parsePathPattern(pathPattern, pathParameters), operationId: operation.operationId, tags: operation.tags });
}
/**
 * @internal
 */
function getRelevantParameters(parameters, refs) {
    const resolvedParameters = parameters
        .map(param => refs.resolveObject(param))
        // Filter cookie parameters
        .filter(param => param.in === 'path' || param.in === 'query' || param.in === 'header');
    return (0, util_1.filterDuplicatesRight)(resolvedParameters, (left, right) => left.name === right.name && left.in === right.in);
}
/**
 * @internal
 */
function parsePathPattern(pathPattern, pathParameters) {
    // Get the innermost curly bracket pairs with non-empty and legal content as placeholders
    const placeholders = pathPattern.match(/{[^/?#{}]+}/g) || [];
    // Get the non-parameter strings as static parts
    const staticParts = pathPattern.split(/{[^/?#{}]+}/);
    const sortedPathParameters = placeholders.map(placeholder => {
        const strippedPlaceholder = placeholder.slice(1, -1);
        const pathParameter = pathParameters.find(param => param.originalName === strippedPlaceholder);
        if (!pathParameter) {
            throw new Error(`Could not find path parameter for placeholder '{${strippedPlaceholder}}'.`);
        }
        return `{${pathParameter.name}}`;
    });
    // Check if all path parameters match placeholders
    const originalParameterNames = pathParameters.map(pathParameter => pathParameter.originalName);
    const missingPlaceholders = originalParameterNames.filter(originalParameterName => !placeholders.includes(`{${originalParameterName}}`));
    if (missingPlaceholders.length) {
        throw new Error(`Could not find placeholder for path parameter(s) ${missingPlaceholders
            .map(placeholder => `'${placeholder}'`)
            .join(', ')}.`);
    }
    return staticParts
        .flatMap((staticPart, index) => {
        const paramName = sortedPathParameters[index];
        return paramName ? [staticPart, paramName] : [staticPart];
    })
        .join('');
}
/**
 * @internal
 */
function parsePathParameters(pathParameters, refs, options) {
    const parsedParameters = parseParameters(pathParameters, refs, options);
    const uniqueNames = (0, unique_naming_1.ensureUniqueNames)(parsedParameters.map(({ originalName }) => originalName), options, {
        reservedWords: [
            'body',
            'queryParameters',
            'headerParameters',
            ...internal_1.reservedJsKeywords
        ]
    });
    return parsedParameters.map((param, i) => {
        param.name = uniqueNames[i];
        return param;
    });
}
/**
 * @internal
 */
function parseParameters(pathParameters, refs, options) {
    return pathParameters.map(param => (Object.assign(Object.assign({}, param), { originalName: param.name, schema: (0, schema_1.parseSchema)(param.schema, refs, options), schemaProperties: {} })));
}
