// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/operation-naming.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nameOperations = nameOperations;
exports.getOperationNameFromPatternAndMethod = getOperationNameFromPatternAndMethod;
const util_1 = require("@sap-cloud-sdk/util");
const extensions_1 = require("../extensions");
/**
 * Give all given operations an initial name.
 * All operations with an extension will be renamed.
 * Operations that don't have a name thereafter, will be named based on their method and path pattern.
 * @param operations - Original information on the operations, that are relevant for parsing.
 * @returns Operation information, where every operation has an operationId.
 * @internal
 */
function nameOperations(operations) {
    const [operationsWithExtensions, operationsWithoutExtensions] = (0, util_1.partition)(operations, ({ operation }) => !!operation[extensions_1.operationNameExtension]);
    validateUniqueExtensions(operationsWithExtensions);
    const renamedOperationsWithExtensions = renameOperations(operationsWithExtensions, renameOperationWithExtension);
    const [namedOperations, unnamedOperations] = (0, util_1.partition)(operationsWithoutExtensions, ({ operation }) => !!operation.operationId);
    const renamedUnnamedOperations = renameOperations(unnamedOperations, ({ pathPattern, method }) => getOperationNameFromPatternAndMethod(pathPattern, method));
    return [
        ...renamedOperationsWithExtensions,
        ...namedOperations,
        ...renamedUnnamedOperations
    ];
}
function validateUniqueExtensions(operations) {
    operations.reduce((uniqueOperationNames, operationInfo) => {
        const operationName = renameOperationWithExtension(operationInfo);
        if (uniqueOperationNames.includes(operationName)) {
            throw new Error(`Operation name '${operationInfo.operation[extensions_1.operationNameExtension]}' provided for '${extensions_1.operationNameExtension}' resolves to '${operationName}' and is not unique.`);
        }
        return [...uniqueOperationNames, operationName];
    }, []);
}
function renameOperationWithExtension({ operation }) {
    const originalOperationName = operation[extensions_1.operationNameExtension];
    if (!originalOperationName) {
        // This should never happen
        throw new Error('Could not rename operation based on extension. Extension is not defined.');
    }
    return (0, util_1.camelCase)(originalOperationName);
}
function renameOperations(operations, renameFn) {
    return operations.map(operationInfo => {
        operationInfo.operation.operationId = renameFn(operationInfo);
        return operationInfo;
    });
}
/**
 * @internal
 */
function getOperationNameFromPatternAndMethod(pattern, method) {
    const [placeholders, pathParts] = (0, util_1.partition)(pattern.split('/'), part => /^\{.+\}$/.test(part));
    const prefix = getSpeakingNameForMethod(method).toLowerCase();
    const base = pathParts.map(part => (0, util_1.pascalCase)(part)).join('');
    const suffix = placeholders.length
        ? 'By' + placeholders.map(part => (0, util_1.pascalCase)(part)).join('And')
        : '';
    return prefix + base + suffix;
}
function getSpeakingNameForMethod(method) {
    return nameMapping[method];
}
const nameMapping = {
    get: 'get',
    post: 'create',
    patch: 'update',
    put: 'update',
    delete: 'delete',
    head: 'getHeadersFor',
    options: 'getOptionsFor',
    trace: 'trace'
};
