// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/refs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenApiDocumentRefs = void 0;
exports.createRefs = createRefs;
const util_1 = require("@sap-cloud-sdk/util");
const swagger_parser_1 = __importDefault(require("@apidevtools/swagger-parser"));
const schema_util_1 = require("../schema-util");
const unique_naming_1 = require("./unique-naming");
const schema_naming_1 = require("./schema-naming");
/**
 * Convenience function to invoke the creation of the OpenApiDocumentRefs builder.
 * @param document - The original OpenAPI document.
 * @param options - Parser options.
 * @returns A promise to the reference representation.
 * @internal
 */
async function createRefs(document, options) {
    return OpenApiDocumentRefs.createRefs(document, options);
}
/**
 * Representation of cross references within a document.
 * Useful when resolving references or getting schema names for referenced schemas.
 * @internal
 */
class OpenApiDocumentRefs {
    /**
     * Create a representation of references within a document.
     * @param document - The original OpenAPI document.
     * @param options - Parser options.
     * @returns A promise to the reference representation.
     */
    static async createRefs(document, options) {
        return new OpenApiDocumentRefs(await swagger_parser_1.default.resolve(document, {
            resolve: { external: options.resolveExternal }
        }), OpenApiDocumentRefs.parseSchemaRefMapping(document, options));
    }
    /**
     * Parse mapping between schema references and their unique names.
     * @param document - The original OpenAPI document.
     * @param options - Parser options.
     * @returns A mapping from schema references to schema naming objects.
     */
    static parseSchemaRefMapping(document, options) {
        var _a;
        const originalNames = Object.keys(((_a = document.components) === null || _a === void 0 ? void 0 : _a.schemas) || {});
        const validSchemaNames = (0, schema_naming_1.ensureValidSchemaNames)(originalNames, options);
        const schemaNames = (0, unique_naming_1.ensureUniqueNames)(validSchemaNames, options, {
            format: util_1.pascalCase,
            separator: '_'
        });
        const fileNames = (0, unique_naming_1.ensureUniqueNames)(schemaNames, options, {
            format: util_1.kebabCase,
            separator: '-',
            reservedWords: ['index']
        });
        return originalNames.reduce((mapping, originalName, i) => (Object.assign(Object.assign({}, mapping), { [`#/components/schemas/${originalName}`]: {
                schemaName: (0, util_1.pascalCase)(options.schemaPrefix) + schemaNames[i],
                fileName: fileNames[i]
            } })), {});
    }
    /**
     * Creates a new instance of `OpenApiDocumentRefs`.
     * @param refs - Object representing the OpenAPI cross references.
     * @param schemaRefMapping - Mapping between schema references and schema naming.
     */
    constructor(refs, schemaRefMapping) {
        this.refs = refs;
        this.schemaRefMapping = schemaRefMapping;
    }
    /**
     * Check whether the given object is a reference object and resolve if necessary.
     * This operates only on the current level and does not resolve the object recursively.
     * @param obj - Object to resolve if necessary.
     * @returns A resolved object.
     */
    resolveObject(obj) {
        return (0, schema_util_1.isReferenceObject)(obj) ? this.refs.get(obj.$ref) : obj;
    }
    /**
     * Parse the type name of a reference object.
     * @param obj - Reference object or reference path to get the type name from.
     * @returns Parsed type name.
     */
    getSchemaNaming(obj) {
        const ref = (0, schema_util_1.isReferenceObject)(obj) ? obj.$ref : obj;
        if (!ref.startsWith('#/components/schemas')) {
            throw new Error(`Could not get schema naming for reference path '${ref}'. Path does not reference a schema.`);
        }
        const schemaNaming = this.schemaRefMapping[ref];
        if (!schemaNaming) {
            throw new Error(`Could not find schema naming for reference path '${ref}'. Schema does not exist.`);
        }
        return schemaNaming;
    }
}
exports.OpenApiDocumentRefs = OpenApiDocumentRefs;
