// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/schema.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseSchema = parseSchema;
exports.parseObjectSchema = parseObjectSchema;
exports.normalizeSchema = normalizeSchema;
exports.parseSchemaProperties = parseSchemaProperties;
const util_1 = require("@sap-cloud-sdk/util");
const schema_util_1 = require("../schema-util");
const type_mapping_1 = require("./type-mapping");
const logger = (0, util_1.createLogger)('openapi-generator');
/**
 * Parse the original schema or reference object to a serializable schema.
 * @param schema - Originally provided schema or reference object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The parsed schema.
 * @internal
 */
function parseSchema(schema, refs, options) {
    var _a, _b, _c, _d;
    if (!schema) {
        logger.verbose("No schema provided, continuing with 'any'.");
        return { type: 'any' };
    }
    if ((0, schema_util_1.isReferenceObject)(schema)) {
        return parseReferenceSchema(schema, refs);
    }
    if (schema.type === 'array') {
        return parseArraySchema(schema, refs, options);
    }
    if ((_a = schema.enum) === null || _a === void 0 ? void 0 : _a.length) {
        return parseEnumSchema(schema, options);
    }
    if (((_b = schema.oneOf) === null || _b === void 0 ? void 0 : _b.length) || schema.discriminator) {
        return parseXOfSchema(schema, refs, 'oneOf', options);
    }
    if ((_c = schema.allOf) === null || _c === void 0 ? void 0 : _c.length) {
        return parseXOfSchema(schema, refs, 'allOf', options);
    }
    if ((_d = schema.anyOf) === null || _d === void 0 ? void 0 : _d.length) {
        return parseXOfSchema(schema, refs, 'anyOf', options);
    }
    // An object schema should be parsed after allOf, anyOf, oneOf.
    // When object.properties are at the same level with anyOf, oneOf, allOf, they should be treated as part of allOf, etc.
    if (schema.type === 'object' ||
        schema.properties ||
        schema.additionalProperties) {
        return parseObjectSchema(schema, refs, options);
    }
    if (schema.not) {
        return {
            not: parseSchema(schema.not, refs, options)
        };
    }
    return {
        type: (0, type_mapping_1.getType)(schema.type)
    };
}
function parseReferenceSchema(schema, refs) {
    return Object.assign(Object.assign({}, schema), refs.getSchemaNaming(schema));
}
/**
 * Parse a schema to an array schema.
 * @param schema - Original schema representing an array.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The recursively parsed array schema.
 */
function parseArraySchema(schema, refs, options) {
    return {
        uniqueItems: schema.uniqueItems,
        items: parseSchema(schema.items, refs, options)
    };
}
/**
 * @internal
 * Parse a schema to an object schema.
 * @param schema - Original schema representing an object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The recursively parsed object schema.
 */
function parseObjectSchema(schema, refs, options) {
    if (schema.discriminator) {
        return parseXOfSchema(schema, refs, 'oneOf', options);
    }
    const properties = parseObjectSchemaProperties(schema, refs, options);
    if (schema.additionalProperties === false) {
        if (!properties.length) {
            throw new Error('Could not parse object schema without neither properties nor additional properties.');
        }
        return { properties };
    }
    const additionalProperties = typeof schema.additionalProperties === 'object' &&
        Object.keys(schema.additionalProperties).length
        ? parseSchema(schema.additionalProperties, refs, options)
        : { type: 'any' };
    return {
        properties,
        additionalProperties
    };
}
/**
 * Parse properties of an object as property schemas.
 * @param schema - Original schema representing an object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The list of parsed property schemas.
 */
function parseObjectSchemaProperties(schema, refs, options) {
    return Object.entries(schema.properties || {}).reduce((props, [propName, propSchema]) => {
        var _a, _b;
        return [
            ...props,
            {
                schema: parseSchema(propSchema, refs, options),
                description: (0, schema_util_1.isReferenceObject)(propSchema)
                    ? undefined
                    : propSchema.description,
                nullable: (_a = (!(0, schema_util_1.isReferenceObject)(propSchema) && propSchema.nullable)) !== null && _a !== void 0 ? _a : false,
                name: propName,
                required: ((_b = schema.required) === null || _b === void 0 ? void 0 : _b.includes(propName)) || false,
                schemaProperties: Object.assign({}, parseSchemaProperties(propSchema))
            }
        ];
    }, []);
}
/**
 * Parse an enum schema.
 * @param schema - Original schema representing an enum.
 * @param options - Options that were set for service generation.
 * @returns The parsed enum schema.
 */
function parseEnumSchema(schema, options) {
    const type = schema.type ? (0, type_mapping_1.getType)(schema.type) : 'string';
    return {
        type,
        enum: (schema.enum || []).map(entry => {
            if (type === 'string' && entry !== null) {
                return getEnumStringValue(String(entry));
            }
            if (entry === null && !schema.nullable) {
                if (options.strictNaming) {
                    throw new Error('null was used as a parameter in an enum, although the schema was not declared as nullable');
                }
                else {
                    logger.warn('null was used as a parameter in an enum, although the schema was not declared as nullable');
                }
            }
            return entry;
        })
    };
}
function getEnumStringValue(input) {
    return `'${input.replace(/'/g, "\\'")}'`;
}
/**
 * Parse a 'oneOf', 'allOf' or 'anyOf' schema.
 * @param schema - Original schema to parse.
 * @param refs - Object representing cross references throughout the document.
 * @param xOf - Key to identify which schema to parse.
 * @param options - Options that were set for service generation.
 * @returns The parsed schema based on the given key.
 */
function parseXOfSchema(schema, refs, xOf, options) {
    const normalizedSchema = normalizeSchema(schema, xOf);
    const xOfSchema = {
        [xOf]: (normalizedSchema[xOf] || []).map(entry => parseSchema(Object.assign(Object.assign({}, entry), { required: [
                // Add required properties from the entry.
                ...('required' in entry && entry.required ? entry.required : []),
                // Add required properties from the top level schema (xOf).
                ...(normalizedSchema.required || [])
            ] }), refs, options))
    };
    if (schema.discriminator && xOf !== 'allOf') {
        return Object.assign(Object.assign({}, xOfSchema), { discriminator: parseDiscriminator(schema, refs, xOf, options) });
    }
    return xOfSchema;
}
function parseDiscriminator(schema, refs, xOf, options) {
    const { discriminator } = schema;
    if (!discriminator) {
        throw new Error('Could not parse discriminator schema without discriminator.');
    }
    const discriminatorMapping = getDiscriminatorMapping(schema, xOf);
    return {
        propertyName: discriminator.propertyName,
        mapping: Object.entries(discriminatorMapping).reduce((mapping, [propertyValue, schemaMapping]) => (Object.assign(Object.assign({}, mapping), { [propertyValue]: parseSchema({ $ref: schemaMapping }, refs, options) })), {})
    };
}
/**
 * @internal
 */
function normalizeSchema(schema, xOf) {
    if (schema.discriminator) {
        return schema;
    }
    if (schema.properties || schema.additionalProperties) {
        logger.info(`Detected schema with ${xOf} and properties in the same level. This was refactored to a schema with ${xOf} only, containing all the properties from the top level.`);
        const _a = schema, _b = xOf, _c = _a[_b], xOfSchema = _c === void 0 ? [] : _c, objectSchema = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
        return { [xOf]: [...xOfSchema, objectSchema] };
    }
    return schema;
}
/**
 * Parse schema properties e.g. 'maxLength', 'minimum', etc.
 * @param schema - Original schema representing a ref or schema object.
 * @returns The parsed schema properties object.
 * @internal
 */
function parseSchemaProperties(schema) {
    if ((0, schema_util_1.isReferenceObject)(schema)) {
        return {};
    }
    const schemaPropertyNames = [
        'deprecated',
        'example',
        'format',
        'default',
        'multipleOf',
        'maximum',
        'minimum',
        'maxLength',
        'minLength',
        'minItems',
        'maxItems',
        'pattern'
    ];
    return schemaPropertyNames.reduce((properties, propertyName) => {
        if (schema[propertyName]) {
            return Object.assign(Object.assign({}, properties), { [propertyName]: schema[propertyName] });
        }
        return properties;
    }, {});
}
function getDiscriminatorMapping(schema, xOf) {
    var _a;
    return (((_a = schema.discriminator) === null || _a === void 0 ? void 0 : _a.mapping) ||
        (schema[xOf] || [])
            .filter(subSchema => (0, schema_util_1.isReferenceObject)(subSchema))
            .reduce((mapping, subSchema) => {
            const originalSchemaName = subSchema.$ref.split('/').reverse()[0];
            return Object.assign(Object.assign({}, mapping), { [originalSchemaName]: subSchema.$ref });
        }, {}));
}
