// @sap-cloud-sdk/openapi-generator@4.4.0/dist/parser/type-mapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getType = getType;
const util_1 = require("@sap-cloud-sdk/util");
const logger = (0, util_1.createLogger)('openapi-generator');
/**
 * Type names based on potential types from the OpenAPI specification.
 * Inspired by: https://github.com/OpenAPITools/openapi-generator/blob/5d27799bade1cfd3b121cda3ddb1599d9aed21e5/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java#L146.
 */
const typeMapping = {
    Set: 'Set',
    set: 'Set',
    Array: 'Array',
    array: 'Array',
    boolean: 'boolean',
    string: 'string',
    int: 'number',
    float: 'number',
    number: 'number',
    long: 'number',
    short: 'number',
    char: 'string',
    double: 'number',
    object: 'Record<string, any>',
    integer: 'number',
    Map: 'any',
    map: 'any',
    date: 'string',
    DateTime: 'string',
    binary: 'any',
    File: 'any',
    file: 'any',
    ByteArray: 'string',
    UUID: 'string',
    URI: 'string',
    Error: 'Error',
    AnyType: 'any',
    any: 'any'
};
/**
 * Get the mapped TypeScript type for the given original OpenAPI type.
 * @param originalType - Original OpenAPI type, to get a mapping for.
 * @returns The mapped TypeScript type.
 * @internal
 */
function getType(originalType) {
    const type = originalType ? typeMapping[originalType] : 'any';
    if (!type) {
        logger.verbose(`Could not map type '${originalType}' to a native type. Using any.`);
        return 'any';
    }
    return type;
}
