// @sap-cloud-sdk/openapi-generator@4.4.0/dist/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = generate;
exports.generateWithParsedOptions = generateWithParsedOptions;
const fs_1 = require("fs");
const path_1 = require("path");
const util_1 = require("@sap-cloud-sdk/util");
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
const file_serializer_1 = require("./file-serializer");
const parser_1 = require("./parser");
const document_converter_1 = require("./document-converter");
const sdk_metadata_1 = require("./sdk-metadata");
const options_1 = require("./options");
const { mkdir } = fs_1.promises;
const logger = (0, util_1.createLogger)('openapi-generator');
/**
 * Main entry point for the OpenAPI client generation.
 * Generates models and API files.
 * @param options - Options to configure generation.
 * @returns A promise to void.
 */
async function generate(options) {
    const parsedOptions = (0, internal_1.parseOptions)(options_1.cliOptions, options);
    if (parsedOptions.verbose) {
        (0, util_1.setLogLevel)('verbose', logger);
    }
    logger.verbose(`Parsed Options: ${JSON.stringify(options, null, 2)}`);
    return generateWithParsedOptions(parsedOptions);
}
/**
 * Main entry point for the OpenAPI client generation.
 * Generates models and API files.
 * @param options - Options to configure generation.
 * @internal
 */
async function generateWithParsedOptions(options) {
    var _a;
    if (!options.input.length || options.outputDir === '') {
        throw new Error('Either input or outputDir were not set.');
    }
    if (options.clearOutputDir) {
        // function rm was added in node version 14 and is the preferred method to use.
        const rm = fs_1.promises.rm || fs_1.promises.rmdir;
        const forceOption = typeof fs_1.promises.rm === 'undefined' ? {} : { force: true };
        await rm(options.outputDir, Object.assign({ recursive: true }, forceOption));
    }
    const inputFilePaths = options.input;
    const optionsPerService = await (0, internal_1.getOptionsPerService)(inputFilePaths, options);
    const tsConfig = await (0, internal_1.tsconfigJson)(options.transpile, options.tsconfig, options.generateESM);
    const promises = inputFilePaths.map(inputFilePath => generateService(inputFilePath, options, optionsPerService[(0, internal_1.getRelPathWithPosixSeparator)(inputFilePath)], tsConfig));
    try {
        const errorMessage = promises.length > 1
            ? 'Some clients could not be generated.'
            : 'Could not generate client.';
        await (0, util_1.finishAll)(promises, errorMessage);
    }
    catch (err) {
        if ((_a = err.message) === null || _a === void 0 ? void 0 : _a.includes('error TS2307')) {
            throw new util_1.ErrorWithCause('Did you forget to install "@sap-cloud-sdk/openapi"?', err);
        }
        throw err;
    }
    finally {
        if (options.optionsPerService) {
            await (0, internal_1.writeOptionsPerService)(options.optionsPerService, optionsPerService, options);
        }
        if (!options.packageJson) {
            logger.info("Finished generation. Don't forget to add @sap-cloud-sdk/openapi to your dependencies.");
        }
    }
}
async function getFileCreationOptions(options) {
    return {
        prettierOptions: await (0, internal_1.readPrettierConfig)(options.prettierConfig),
        overwrite: options.overwrite,
        generateESM: options.generateESM
    };
}
/**
 * Generate sources.
 * @param serviceDir - Directory to generate the service to.
 * @param openApiDocument - Parsed service.
 * @param inputFilePath - Path to the input file.
 * @param tsConfig - File content for the `tsconfig.json`.
 * @param options - Generation options.
 */
async function generateSources(serviceDir, openApiDocument, inputFilePath, tsConfig, options) {
    await mkdir(serviceDir, { recursive: true });
    await generateMandatorySources(serviceDir, openApiDocument, options);
    if (options.metadata) {
        await generateMetadata(openApiDocument, inputFilePath, options);
    }
    if (options.packageJson) {
        await generatePackageJson(serviceDir, openApiDocument, options);
    }
    if (options.include) {
        await (0, internal_1.copyFiles)(options.include, serviceDir, options.overwrite);
    }
    if (tsConfig) {
        await (0, internal_1.createFile)(serviceDir, 'tsconfig.json', tsConfig, await getFileCreationOptions(options));
        const transpileOptions = {
            compilerOptions: await (0, internal_1.readCompilerOptions)(serviceDir),
            createFileOptions: await getFileCreationOptions(options)
        };
        await (0, internal_1.transpileDirectory)(serviceDir, transpileOptions);
    }
    if (options.readme) {
        await generateReadme(serviceDir, openApiDocument, options);
    }
}
async function generateMandatorySources(serviceDir, openApiDocument, options) {
    const createFileOptions = await getFileCreationOptions(options);
    if (openApiDocument.schemas.length) {
        const schemaDir = (0, path_1.resolve)(serviceDir, 'schema');
        await createSchemaFiles(schemaDir, openApiDocument, createFileOptions);
        await (0, internal_1.createFile)(schemaDir, 'index.ts', (0, file_serializer_1.schemaIndexFile)(openApiDocument, createFileOptions), createFileOptions);
    }
    await createApis(serviceDir, openApiDocument, createFileOptions);
    await (0, internal_1.createFile)(serviceDir, 'index.ts', (0, file_serializer_1.apiIndexFile)(openApiDocument, createFileOptions), createFileOptions);
}
async function createApis(serviceDir, openApiDocument, options) {
    await Promise.all(openApiDocument.apis.map(api => (0, internal_1.createFile)(serviceDir, `${(0, util_1.kebabCase)(api.name)}.ts`, (0, file_serializer_1.apiFile)(api, openApiDocument.serviceName, options, openApiDocument.serviceOptions.basePath), options)));
}
async function createSchemaFiles(dir, openApiDocument, createFileOptions) {
    await mkdir(dir, { recursive: true });
    await Promise.all(openApiDocument.schemas.map(schema => (0, internal_1.createFile)(dir, `${schema.fileName}.ts`, (0, file_serializer_1.schemaFile)(schema, createFileOptions), createFileOptions)));
}
/**
 * Generates an OpenAPI service from a file.
 * @param inputFilePath - The file path where the service to generate is located.
 * @param options - Options to configure generation.
 * @param serviceOptions - Service options as defined in the options per service.
 * @param tsConfig - File content for the `tsconfig.json`.
 */
async function generateService(inputFilePath, options, serviceOptions, tsConfig) {
    const openApiDocument = await (0, document_converter_1.convertOpenApiSpec)(inputFilePath);
    const parsedOpenApiDocument = await (0, parser_1.parseOpenApiDocument)(openApiDocument, serviceOptions, {
        strictNaming: !options.skipValidation,
        schemaPrefix: options.schemaPrefix,
        resolveExternal: options.resolveExternal
    });
    const serviceDir = (0, path_1.resolve)(options.outputDir, serviceOptions.directoryName);
    await generateSources(serviceDir, parsedOpenApiDocument, inputFilePath, tsConfig, options);
    logger.info(`Successfully generated client for '${inputFilePath}'`);
}
async function generateReadme(serviceDir, openApiDocument, options) {
    logger.verbose(`Generating readme in ${serviceDir}.`);
    return (0, internal_1.createFile)(serviceDir, 'README.md', (0, file_serializer_1.readme)(openApiDocument), await getFileCreationOptions(options));
}
async function generateMetadata(openApiDocument, inputFilePath, options) {
    const { name: inputFileName, dir: inputDirPath } = (0, path_1.parse)(inputFilePath);
    const { clientFileName } = (0, internal_1.getSdkMetadataFileNames)(inputFileName);
    const metadataDir = (0, path_1.resolve)(inputDirPath, 'sdk-metadata');
    await mkdir(metadataDir, { recursive: true });
    logger.verbose(`Generating client metadata ${clientFileName}...`);
    const createFileOptions = await getFileCreationOptions(options);
    const clientFile = (0, internal_1.createFile)(metadataDir, clientFileName, JSON.stringify(await (0, sdk_metadata_1.sdkMetadata)(openApiDocument), null, 2), createFileOptions);
    return clientFile;
}
async function generatePackageJson(serviceDir, openApiDocument, options) {
    logger.verbose(`Generating package.json in ${serviceDir}.`);
    const createFileOptions = await getFileCreationOptions(options);
    await (0, internal_1.createFile)(serviceDir, 'package.json', (0, file_serializer_1.packageJson)({
        npmPackageName: openApiDocument.serviceOptions.packageName,
        description: (0, internal_1.packageDescription)(openApiDocument.serviceName),
        sdkVersion: await (0, internal_1.getSdkVersion)(),
        generateESM: options.generateESM
    }), createFileOptions);
}
