// @sap-cloud-sdk/openapi-generator@4.4.0/dist/schema-util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectRefs = collectRefs;
exports.getUniqueRefs = getUniqueRefs;
exports.isReferenceObject = isReferenceObject;
exports.isArraySchema = isArraySchema;
exports.isObjectSchema = isObjectSchema;
exports.isEnumSchema = isEnumSchema;
exports.isOneOfSchema = isOneOfSchema;
exports.isAllOfSchema = isAllOfSchema;
exports.isAnyOfSchema = isAnyOfSchema;
exports.isNotSchema = isNotSchema;
exports.getSchemaNamingFromRef = getSchemaNamingFromRef;
exports.getSchemaPropertiesDocumentation = getSchemaPropertiesDocumentation;
const util_1 = require("@sap-cloud-sdk/util");
/**
 * Collect all unique reference schemas within the given schemas.
 * @param schemas - Parsed schemas to retrieve all references for.
 * @returns Returns a list of unique reference schemas within the given schemas.
 * @internal
 */
function collectRefs(...schemas) {
    return getUniqueRefs((0, util_1.flat)(schemas.map(schema => collectAllRefs(schema))));
}
/**
 * Collect all reference schemas within a schema.
 * The resulting list of references might contain duplicates.
 * @param schema - Parsed schema to retrieve all references for.
 * @returns Returns a list of reference schemas within a schema. Might contain duplicates.
 */
function collectAllRefs(schema) {
    if (!schema) {
        return [];
    }
    if (isReferenceObject(schema)) {
        return [schema];
    }
    return Object.values(schema)
        .filter(value => typeof value === 'object')
        .reduce((refs, value) => [...refs, ...collectAllRefs(value)], []);
}
/**
 * Reduce a list of reference schemas to a list of unique reference schemas, based on $ref.
 * @param refs - List of reference schemas.
 * @returns List of unique reference schemas.
 * @internal
 */
function getUniqueRefs(refs) {
    return refs.reduce((uniqueRefs, collectedRef) => {
        if (!uniqueRefs.some(ref => ref.$ref === collectedRef.$ref)) {
            uniqueRefs.push(collectedRef);
        }
        return uniqueRefs;
    }, []);
}
/**
 * Type guard to check whether an object is of type `OpenAPIV3.ReferenceObject`.
 * @param obj - Object to check.
 * @returns `true` if the object is a reference object, `false` otherwise.
 * @internal
 */
function isReferenceObject(obj) {
    return !!(obj === null || obj === void 0 ? void 0 : obj.$ref);
}
/**
 * Type guard to check whether an object is of type `OpenApiArraySchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is an array schema, `false` otherwise.
 * @internal
 */
function isArraySchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.items;
}
/**
 * Type guard to check whether an object is of type `OpenApiObjectSchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is an object schema, `false` otherwise.
 * @internal
 */
function isObjectSchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.properties;
}
/**
 * Type guard to check whether an object is of type `OpenApiEnumSchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is an enum schema, `false` otherwise.
 * @internal
 */
function isEnumSchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.enum;
}
/**
 * Type guard to check whether an object is of type `OpenApiOneOfSchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is an oneOf schema, `false` otherwise.
 * @internal
 */
function isOneOfSchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.oneOf;
}
/**
 * Type guard to check whether an object is of type `OpenApiAllOfSchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is an allOf schema, `false` otherwise.
 * @internal
 */
function isAllOfSchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.allOf;
}
/**
 * Type guard to check whether an object is of type `OpenApiAnyOfSchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is an anyOf schema, `false` otherwise.
 * @internal
 */
function isAnyOfSchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.anyOf;
}
/**
 * Type guard to check whether an object is of type `OpenApiNotSchema`.
 * @param obj - Object to check.
 * @returns `true` if the object is a not schema, `false` otherwise.
 * @internal
 */
function isNotSchema(obj) {
    return obj === null || obj === void 0 ? void 0 : obj.not;
}
/**
 * Parse the type name of a reference object.
 * @param obj - Reference object to get the type name from.
 * @param schemaRefMapping - Mapping between reference paths and schema names.
 * @returns Parsed type name.
 * @internal
 */
function getSchemaNamingFromRef(obj, schemaRefMapping) {
    const ref = isReferenceObject(obj) ? obj.$ref : obj;
    const schemaNaming = schemaRefMapping[ref];
    if (!schemaNaming) {
        throw new Error(`Could not find schema naming for reference path '${ref}'. Schema does not exist.`);
    }
    return schemaNaming;
}
/**
 * @internal
 */
function getSchemaPropertiesDocumentation(schemaProperties) {
    return Object.entries(schemaProperties || []).map(([propertyName, value]) => {
        if (propertyName === 'deprecated') {
            return '@deprecated';
        }
        if (propertyName === 'example') {
            return `@example ${JSON.stringify(schemaProperties === null || schemaProperties === void 0 ? void 0 : schemaProperties.example, null, 2)}`;
        }
        return `${(0, util_1.titleFormat)(propertyName)}: ${JSON.stringify(value, null, 2)}.`;
    });
}
