// @sap-cloud-sdk/openapi-generator@4.4.0/dist/file-serializer/api-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiFile = apiFile;
exports.apiDocumentation = apiDocumentation;
const util_1 = require("@sap-cloud-sdk/util");
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
const schema_util_1 = require("../schema-util");
const operation_1 = require("./operation");
/**
 * Serialize an API representation to a string representing the resulting API file.
 * @param api - Representation of an API.
 * @param serviceName - Service name for which the API is created.
 * @param options - Options to configure the file creation.
 * @param basePath - Custom base path for the API.
 * @returns The serialized API file contents.
 * @internal
 */
function apiFile(api, serviceName, options, basePath) {
    const imports = (0, internal_1.serializeImports)(getImports(api, options));
    const apiDoc = apiDocumentation(api, serviceName);
    const apiContent = (0, util_1.codeBlock) `
export const ${api.name} = {
  _defaultBasePath: ${basePath ? `'${basePath}'` : undefined},
  ${api.operations.map(operation => (0, operation_1.serializeOperation)(operation, api.name)).join(',\n')}
};
`;
    return [imports, apiDoc, apiContent].join(util_1.unixEOL);
}
/**
 * Get the unique reference schemas for all request body types in the given operation list.
 * @param operations - The given operation list.
 * @returns The list of unique referenced body type schemas.
 */
function collectRefsFromOperations(operations) {
    return (0, schema_util_1.getUniqueRefs)(operations.reduce((referenceTypes, operation) => {
        var _a;
        return [
            ...referenceTypes,
            ...(0, schema_util_1.collectRefs)(...[
                (_a = operation.requestBody) === null || _a === void 0 ? void 0 : _a.schema,
                operation.response,
                ...operation.queryParameters.map(({ schema }) => schema)
            ])
        ];
    }, []));
}
function getImports(api, options) {
    const refs = collectRefsFromOperations(api.operations).map(requestBodyType => requestBodyType.schemaName);
    const refImports = {
        names: refs,
        moduleIdentifier: (options === null || options === void 0 ? void 0 : options.generateESM) ? './schema/index.js' : './schema',
        typeOnly: true
    };
    const openApiImports = {
        names: ['OpenApiRequestBuilder'],
        moduleIdentifier: '@sap-cloud-sdk/openapi'
    };
    return [openApiImports, refImports];
}
/**
 * @internal
 */
function apiDocumentation(api, serviceName) {
    return (0, util_1.documentationBlock) `
  Representation of the '${api.name}'.
  This API is part of the '${serviceName}' service.`;
}
