// @sap-cloud-sdk/openapi-generator@4.4.0/dist/file-serializer/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeOperation = serializeOperation;
exports.operationDocumentation = operationDocumentation;
const util_1 = require("@sap-cloud-sdk/util");
const schema_1 = require("./schema");
/**
 * Serialize an operation to a string.
 * @param operation - Operation to serialize.
 * @param apiName - Name of the API the operation is part of.
 * @returns The operation as a string.
 * @internal
 */
function serializeOperation(operation, apiName) {
    const requestBuilderParams = [
        `'${operation.method}'`,
        `"${operation.pathPattern}"`
    ];
    const bodyAndQueryParams = serializeParamsForRequestBuilder(operation);
    if (bodyAndQueryParams) {
        requestBuilderParams.push(bodyAndQueryParams);
    }
    else {
        requestBuilderParams.push('{}');
    }
    const responseType = (0, schema_1.serializeSchema)(operation.response);
    return (0, util_1.codeBlock) `
${operationDocumentation(operation)}
${operation.operationId}: (${serializeOperationSignature(operation)}) => new OpenApiRequestBuilder<${responseType}>(
  ${requestBuilderParams.join(',\n')},
  ${apiName}._defaultBasePath
)`;
}
function serializeOperationSignature(operation) {
    const pathParams = serializePathParamsForSignature(operation);
    const requestBodyParam = serializeRequestBodyParamForSignature(operation);
    const allOptionalHeaders = operation.headerParameters.every(param => !param.required);
    const allOptionalQuery = operation.queryParameters.every(param => !param.required);
    const headerParams = serializeParamsForSignature(operation, 'headerParameters', allOptionalHeaders);
    const queryParams = serializeParamsForSignature(operation, 'queryParameters', allOptionalHeaders && allOptionalQuery);
    return [pathParams, requestBodyParam, queryParams, headerParams]
        .filter(params => params)
        .join(', ');
}
function serializePathParamsForSignature(operation) {
    if (operation.pathParameters.length) {
        return operation.pathParameters
            .map(param => `${param.name}: string`)
            .join(', ');
    }
}
function serializeRequestBodyParamForSignature(operation) {
    if (operation.requestBody) {
        return `body: ${(0, schema_1.serializeSchema)(operation.requestBody.schema)}${operation.requestBody.required ? '' : ' | undefined'}`;
    }
}
function serializeParamsForSignature(operation, paramType, isAllOptional) {
    const parameters = operation[paramType];
    if (parameters.length) {
        const paramsString = parameters
            .map(param => `'${param.name}'${param.required ? '' : '?'}: ${(0, schema_1.serializeSchema)(param.schema)}`)
            .join(', ');
        return `${paramType}${isAllOptional ? '?' : ''}: {${paramsString}}`;
    }
}
function serializeParamsForRequestBuilder(operation) {
    const params = [];
    if (operation.pathParameters.length) {
        const pathParamStr = `pathParameters: { ${operation.pathParameters
            .map(param => param.name)
            .join(', ')} }`;
        params.push(pathParamStr);
    }
    if (operation.requestBody) {
        params.push('body');
    }
    if (operation.queryParameters.length) {
        params.push('queryParameters');
    }
    if (operation.headerParameters.length) {
        params.push('headerParameters');
    }
    if (params.length) {
        return (0, util_1.codeBlock) `{
      ${params.join(',\n')}
    }`;
    }
}
/**
 * @internal
 */
function operationDocumentation(operation) {
    var _a;
    const signature = [];
    if (operation.pathParameters.length) {
        signature.push(...getSignatureOfPathParameters(operation.pathParameters));
    }
    if (operation.requestBody) {
        signature.push(getSignatureOfBody(operation.requestBody));
    }
    if (operation.queryParameters.length) {
        signature.push(`@param queryParameters - Object containing the following keys: ${operation.queryParameters
            .map(param => `${param.name}`)
            .join(', ')}.`);
    }
    if ((_a = operation.headerParameters) === null || _a === void 0 ? void 0 : _a.length) {
        signature.push(`@param headerParameters - Object containing the following keys: ${operation.headerParameters
            .map(param => `${param.name}`)
            .join(', ')}.`);
    }
    signature.push('@returns The request builder, use the `execute()` method to trigger the request.');
    const lines = [getOperationDescriptionText(operation), ...signature];
    return (0, util_1.documentationBlock) `${lines.join(util_1.unixEOL)}`;
}
function getSignatureOfPathParameters(parameters) {
    return parameters.map(parameter => `@param ${parameter.name} - ${parameter.description || 'Path parameter.'}`);
}
function getSignatureOfBody(body) {
    return `@param body - ${body.description || 'Request body.'}`;
}
function getOperationDescriptionText(operation) {
    if (operation.description) {
        return operation.description;
    }
    return `Create a request builder for execution of ${operation.method} requests to the '${operation.pathPattern}' endpoint.`;
}
