// @sap-cloud-sdk/openapi-generator@4.4.0/dist/file-serializer/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeSchema = serializeSchema;
exports.schemaPropertyDocumentation = schemaPropertyDocumentation;
const util_1 = require("@sap-cloud-sdk/util");
const parser_1 = require("../parser");
const schema_util_1 = require("../schema-util");
/**
 * Serialize a schema.
 * @param schema - Parsed schema to be serialized.
 * @returns The serialized schema
 * @internal
 */
function serializeSchema(schema) {
    if ((0, schema_util_1.isReferenceObject)(schema)) {
        return schema.schemaName;
    }
    if ((0, schema_util_1.isArraySchema)(schema)) {
        const type = serializeSchema(schema.items);
        return schema.uniqueItems
            ? `Set<${type}>`
            : ['properties', 'allOf', 'oneOf', 'anyOf', 'enum'].some(prop => prop in schema.items)
                ? `(${type})[]`
                : `${type}[]`;
    }
    if ((0, schema_util_1.isObjectSchema)(schema)) {
        return serializeObjectSchema(schema);
    }
    if ((0, schema_util_1.isEnumSchema)(schema)) {
        return schema.enum.join(' | ');
    }
    if ((0, schema_util_1.isOneOfSchema)(schema)) {
        if (schema.discriminator) {
            return serializeXOfSchemaWithDiscriminator(schema);
        }
        return schema.oneOf.map(type => serializeSchema(type)).join(' | ');
    }
    if ((0, schema_util_1.isAllOfSchema)(schema)) {
        return schema.allOf.map(type => serializeSchema(type)).join(' & ');
    }
    if ((0, schema_util_1.isAnyOfSchema)(schema)) {
        if (schema.discriminator) {
            return serializeXOfSchemaWithDiscriminator(schema);
        }
        return schema.anyOf.map(type => serializeSchema(type)).join(' | ');
    }
    if ((0, schema_util_1.isNotSchema)(schema)) {
        return (0, util_1.codeBlock) `any`;
    }
    return (0, parser_1.getType)(schema.type);
}
function serializeObjectSchema(schema) {
    if (!schema.properties.length &&
        typeof schema.additionalProperties !== 'object') {
        return 'Record<string, any>';
    }
    const types = [];
    if (schema.properties.length) {
        types.push(serializeObjectSchemaForProperties(schema.properties));
    }
    if (schema.additionalProperties) {
        types.push(`Record<string, ${serializeSchema(schema.additionalProperties)}>`);
    }
    return types.join(' & ');
}
function serializeObjectSchemaForProperties(properties) {
    return (0, util_1.codeBlock) `{
      ${properties
        .map(property => serializePropertyWithDocumentation(property))
        .join(util_1.unixEOL)}
    }`;
}
function serializeProperty(name, required, nullable, type) {
    return `'${name}'${required ? '' : '?'}: ${type}${nullable ? ' | null' : ''};`;
}
function serializePropertyWithDocumentation(property) {
    const documentation = schemaPropertyDocumentation(property);
    const serialized = serializeProperty(property.name, property.required, property.nullable, serializeSchema(property.schema));
    if (documentation) {
        return [documentation, serialized].join(util_1.unixEOL);
    }
    return serialized;
}
/**
 * @internal
 */
function schemaPropertyDocumentation(schema) {
    const signature = [];
    if (schema.description) {
        signature.push(schema.description);
    }
    signature.push(...(0, schema_util_1.getSchemaPropertiesDocumentation)(schema.schemaProperties));
    return (0, util_1.documentationBlock) `${signature.join(util_1.unixEOL)}`;
}
function serializeXOfSchemaWithDiscriminator(schema) {
    const { discriminator } = schema;
    if (!discriminator) {
        throw new Error('Could not serialize discriminator schema without discriminator.');
    }
    return Object.entries(discriminator.mapping)
        .map(([propertyValue, mappedSchema]) => `({ ${discriminator.propertyName}: '${propertyValue}' } & ${serializeSchema(mappedSchema)})`)
        .join(' | ');
}
