// @sap-cloud-sdk/openapi-generator@4.4.0/dist/file-serializer/schema-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemaFile = schemaFile;
exports.schemaDocumentation = schemaDocumentation;
const util_1 = require("@sap-cloud-sdk/util");
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
const schema_util_1 = require("../schema-util");
const schema_1 = require("./schema");
/**
 * Serialize a schema representation to a string representing the according schema file contents.
 * @param namedSchema - A named schema.
 * @returns The serialized schema file contents.
 * @internal
 */
function schemaFile(namedSchema, options) {
    const imports = (0, internal_1.serializeImports)(getImports(namedSchema, options));
    return (0, util_1.codeBlock) `    
    ${imports}
    ${schemaDocumentation(namedSchema)}
    export type ${namedSchema.schemaName} = ${(0, schema_1.serializeSchema)(namedSchema.schema)}${namedSchema.nullable ? ' | null' : ''};
  `;
}
function getImports(namedSchema, options) {
    return (0, schema_util_1.collectRefs)(namedSchema.schema)
        .filter(ref => ref.schemaName !== namedSchema.schemaName)
        .map(ref => ({
        names: [ref.schemaName],
        typeOnly: true,
        moduleIdentifier: (options === null || options === void 0 ? void 0 : options.generateESM)
            ? `./${ref.fileName}.js`
            : `./${ref.fileName}`
    }));
}
/**
 * @internal
 */
function schemaDocumentation(schema) {
    const lines = [
        schema.description ||
            `Representation of the '${schema.schemaName}' schema.`,
        ...(0, schema_util_1.getSchemaPropertiesDocumentation)(schema.schemaProperties)
    ];
    return (0, util_1.documentationBlock) `${lines.join(util_1.unixEOL)}`;
}
