// @sap-cloud-sdk/openapi-generator@4.4.0/dist/sdk-metadata/code-sample-util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMainApi = getMainApi;
exports.getApiWithMaxOperations = getApiWithMaxOperations;
exports.getMainOperation = getMainOperation;
exports.getGetAllOperation = getGetAllOperation;
exports.getGetOperation = getGetOperation;
exports.getAnyOperationWithoutParams = getAnyOperationWithoutParams;
exports.getOperationParamCode = getOperationParamCode;
const internal_1 = require("@sap-cloud-sdk/generator-common/internal");
const util_1 = require("@sap-cloud-sdk/util");
/**
 * @internal
 */
function getMainApi(serviceName, apis) {
    return ((0, internal_1.getLevenshteinClosest)(serviceName, apis, x => x.name) ||
        getApiWithMaxOperations(apis));
}
/**
 * @internal
 */
function getApiWithMaxOperations(apis) {
    const sortedByOperationsLength = apis.sort((a, b) => a.operations.length > b.operations.length ? -1 : 1);
    const withGetAll = sortedByOperationsLength.find(api => api.operations.find(operation => operation.operationId.toLocaleLowerCase('getall')));
    if (withGetAll) {
        return withGetAll;
    }
    // return the one with most methods
    return sortedByOperationsLength[0];
}
/**
 * @internal
 */
function getMainOperation(api) {
    if (api.operations.length === 1) {
        return api.operations[0];
    }
    return ((0, internal_1.getLevenshteinClosest)(api.name, api.operations, op => op.operationId) ||
        getGetAllOperation(api.operations) ||
        getGetOperation(api.operations) ||
        getAnyOperationWithoutParams(api.operations) ||
        api.operations[0]);
}
/**
 * @internal
 */
function getGetAllOperation(operations) {
    return operations.find(operation => operation.operationId.toLowerCase() === 'getall');
}
/**
 * @internal
 */
function getGetOperation(operations) {
    const getOperations = operations.filter(operation => operation.method === 'get');
    // check if get operation without parameters exists
    // else return operation with min parameters
    if (getOperations.length > 0) {
        const getOperationsWithoutParams = getOperations.filter(getOperation => getOperation.pathParameters.length === 0 &&
            getOperation.queryParameters.length === 0);
        if (getOperationsWithoutParams.length > 0) {
            return getOperationsWithoutParams[0];
        }
        return getOperations[0];
    }
    return undefined;
}
/**
 * @internal
 */
function getAnyOperationWithoutParams(operations) {
    return operations.find(operation => operation.pathParameters.length === 0 &&
        operation.queryParameters.length === 0);
}
/**
 * @internal
 */
function getOperationParamCode(operation) {
    var _a;
    const paramSignature = [];
    if ((_a = operation.pathParameters) === null || _a === void 0 ? void 0 : _a.length) {
        paramSignature.push(operation.pathParameters.map(pathParam => pathParam.name).join(', '));
    }
    if (operation.requestBody) {
        let reqBody = 'undefined';
        if (operation.requestBody.required) {
            reqBody = 'myRequestBody';
        }
        paramSignature.push(reqBody);
    }
    const formatParams = (params) => params
        .filter(param => param.required)
        .map(param => `${param.name}: 'my${(0, util_1.pascalCase)(param.name)}'`)
        .join(', ');
    if (operation.queryParameters) {
        const queryParams = formatParams(operation.queryParameters);
        if (queryParams) {
            paramSignature.push(`{ ${queryParams} }`);
        }
    }
    if (operation.headerParameters) {
        const headerParams = formatParams(operation.headerParameters);
        if (headerParams) {
            paramSignature.push(`{ ${headerParams} }`);
        }
    }
    return (0, util_1.codeBlock) `${paramSignature.join(', ')}`;
}
